### ISSUE TITLE
Alamofire 5.6.1 crash: ServerTrustEvaluation.swift - Line 611

### ISSUE THREAD

**Issue by Kaspik** [Alamofire/Alamofire#3612]
* [x] I've read, understood, and done my best to follow the [*CONTRIBUTING guidelines](https://github.com/Alamofire/Alamofire/blob/master/CONTRIBUTING.md).

## What happened
Really small number of users are getting crash on app start when Alamofire tries to evaluate certificates. This wasn't happening on 5.5 version but started happening when we updated to 5.6.1 release. For most of the users this is one-time crash, for some it was multiple-times crash.

- Was there a change that we should be aware of during the migration?
- Is there anything else / part of the code I can provide you with? Stack mostly seems related only to Alamofire and not anything in our project

## Stack
<img width="476" alt="Screenshot 2022-05-16 at 20 09 26" src="https://user-images.githubusercontent.com/7204168/168655921-1c875b48-249c-4d87-8356-d0b96da685b7.png">
<img width="1278" alt="Screenshot 2022-05-16 at 20 09 32" src="https://user-images.githubusercontent.com/7204168/168655930-34b48fc1-7a64-495b-9aa5-ec159e84f1e2.png">

Text version:
<details>

```
crash_info_entry_0
Metadata allocator corruption: allocation is NULL. curState: {0x0, 10312} - curStateReRead: {0x0, 10312} - newState: {0x20, 10280} - allocatedNewPage: false - requested size: 32 - sizeWithHeader: 32 - alignment: 8 - Tag: 14

crash_info_entry_1
abort() called
```
```
Crashed: org.alamofire.session.rootQueue
0  libsystem_kernel.dylib         0x6bbc __pthread_kill + 8
1  libsystem_pthread.dylib        0xd854 pthread_kill + 208
2  libsystem_c.dylib              0x1f6a4 abort + 124
3  libswiftCore.dylib             0x2ed01c swift::warningv(unsigned int, char const*, char*) + 138
4  libswiftCore.dylib             0x2f6698 swift::MetadataAllocator::Allocate(unsigned long, unsigned long) + 516
5  libswiftCore.dylib             0x2f71f4 _swift_getGenericMetadata(swift::MetadataRequest, void const* const*, swift::TargetTypeContextDescriptor<swift::InProcess> const*) + 1004
6  libswiftCore.dylib             0x2dad68 __swift_instantiateCanonicalPrespecializedGenericMetadata + 32
7  libswiftFoundation.dylib       0x7a30 specialized _arrayConditionalCast<A, B>(_:) + 48
8  libswiftFoundation.dylib       0x1660c static Array._conditionallyBridgeFromObjectiveC(_:result:) + 80
9  ClassDojo                      0xf54fa0 AlamofireExtension<A>.certificates.getter + 611 (ServerTrustEvaluation.swift:611)
10 ClassDojo                      0xf54714 PinnedCertificatesTrustEvaluator.evaluate(_:forHost:) + 629 (ServerTrustEvaluation.swift:629)
11 ClassDojo                      0xf55110 protocol witness for ServerTrustEvaluating.evaluate(_:forHost:) in conformance PinnedCertificatesTrustEvaluator + 4354347280 (<compiler-generated>:4354347280)
12 ClassDojo                      0xf6ba6c SessionDelegate.attemptServerTrustAuthentication(with:) + 134 (SessionDelegate.swift:134)
13 ClassDojo                      0xf6e50c specialized SessionDelegate.urlSession(_:task:didReceive:completionHandler:) + 99 (<compiler-generated>:99)
14 ClassDojo                      0xf6d224 @objc SessionDelegate.urlSession(_:task:didReceive:completionHandler:) + 4354445860
15 CFNetwork                      0x7d898 _CFHostIsDomainTopLevelForCertificatePolicy + 10480
16 Foundation                     0x4e26c __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 16
17 Foundation                     0x5f920 -[NSBlockOperation main] + 100
18 Foundation                     0x39a8c __NSOPERATION_IS_INVOKING_MAIN__ + 20
19 Foundation                     0x49f20 -[NSOperation start] + 764
20 Foundation                     0x4d368 __NSOPERATIONQUEUE_IS_STARTING_AN_OPERATION__ + 20
21 Foundation                     0x5ab7c __NSOQSchedule_f + 180
22 libdispatch.dylib              0x46048 _dispatch_block_async_invoke2 + 104
23 libdispatch.dylib              0x64094 _dispatch_client_callout + 16
24 libdispatch.dylib              0x3f4a4 _dispatch_lane_serial_drain$VARIANT$armv81 + 600
25 libdispatch.dylib              0x3ff78 _dispatch_lane_invoke$VARIANT$armv81 + 440
26 libdispatch.dylib              0x3f380 _dispatch_lane_serial_drain$VARIANT$armv81 + 308
27 libdispatch.dylib              0x3ff44 _dispatch_lane_invoke$VARIANT$armv81 + 388
28 libdispatch.dylib              0x498e0 _dispatch_workloop_worker_thread + 608
29 libsystem_pthread.dylib        0x1e10 _pthread_wqthread + 284
30 libsystem_pthread.dylib        0x193c start_wqthread + 8
```
```
com.apple.main-thread
0  libsystem_kernel.dylib         0xaac mach_msg_trap + 8
1  libsystem_kernel.dylib         0x107c mach_msg + 72
2  CoreFoundation                 0x6768 __CFRunLoopServiceMachPort + 368
3  CoreFoundation                 0xaa70 __CFRunLoopRun + 1160
4  CoreFoundation                 0x1dc30 CFRunLoopRunSpecific + 572
5  GraphicsServices               0x1988 GSEventRunModal + 160
6  UIKitCore                      0x4e5c50 -[UIApplication _run] + 1080
7  UIKitCore                      0x27f3d0 UIApplicationMain + 336
8  libswiftUIKit.dylib            0x27ee4 UIApplicationMain(_:_:_:_:) + 100
9  ClassDojo                      0x564c main + 4338292300 (MetadataKeys.swift:4338292300)
10 ???                            0x105ed83d0 (Missing)
```
```
trust
0  libsystem_kernel.dylib         0xaac mach_msg_trap + 8
1  libsystem_kernel.dylib         0x107c mach_msg + 72
2  libdispatch.dylib              0x36104 _dispatch_mach_send_and_wait_for_reply + 524
3  libdispatch.dylib              0x36488 dispatch_mach_send_with_result_and_wait_for_reply$VARIANT$armv81 + 56
4  libxpc.dylib                   0xcb10 xpc_connection_send_message_with_reply_sync + 236
5  Security                       0x7c08 securityd_message_with_reply_sync + 96
6  Security                       0x12310 securityd_send_sync_and_do + 80
7  Security                       0x1ea54 __SecTrustEvaluateIfNecessary_block_invoke_3 + 452
8  Security                       0xac4c SecOSStatusWith + 48
9  Security                       0xda2e0 __SecTrustEvaluateIfNecessary_block_invoke_2 + 88
10 libsystem_trace.dylib          0x6d94 _os_activity_initiate_impl + 56
11 Security                       0xda17c __SecTrustEvaluateIfNecessary_block_invoke + 264
12 libdispatch.dylib              0x64094 _dispatch_client_callout + 16
13 libdispatch.dylib              0x46464 _dispatch_lane_barrier_sync_invoke_and_complete + 52
14 Security                       0xd9e7c SecTrustEvaluateIfNecessary + 156
15 Security                       0xd9bc SecTrustEvaluateInternal + 44
16 Security                       0x1f59c SecTrustEvaluateWithError + 36
17 ClassDojo                      0xf56208 AlamofireExtension<A>.evaluate() + 557 (ServerTrustEvaluation.swift:557)
18 ClassDojo                      0xf530bc AlamofireExtension<A>.performValidation(forHost:) + 515 (ServerTrustEvaluation.swift:515)
19 ClassDojo                      0xf546f4 PinnedCertificatesTrustEvaluator.evaluate(_:forHost:) + 271 (ServerTrustEvaluation.swift:271)
20 ClassDojo                      0xf55110 protocol witness for ServerTrustEvaluating.evaluate(_:forHost:) in conformance PinnedCertificatesTrustEvaluator + 4354347280 (<compiler-generated>:4354347280)
21 ClassDojo                      0xf6ba6c SessionDelegate.attemptServerTrustAuthentication(with:) + 134 (SessionDelegate.swift:134)
22 ClassDojo                      0xf6e50c specialized SessionDelegate.urlSession(_:task:didReceive:completionHandler:) + 99 (<compiler-generated>:99)
23 ClassDojo                      0xf6d224 @objc SessionDelegate.urlSession(_:task:didReceive:completionHandler:) + 4354445860
24 CFNetwork                      0x7d898 _CFHostIsDomainTopLevelForCertificatePolicy + 10480
25 Foundation                     0x4e26c __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 16
26 Foundation                     0x5f920 -[NSBlockOperation main] + 100
27 Foundation                     0x39a8c __NSOPERATION_IS_INVOKING_MAIN__ + 20
28 Foundation                     0x49f20 -[NSOperation start] + 764
29 Foundation                     0x4d368 __NSOPERATIONQUEUE_IS_STARTING_AN_OPERATION__ + 20
30 Foundation                     0x5ab7c __NSOQSchedule_f + 180
31 libdispatch.dylib              0x46048 _dispatch_block_async_invoke2 + 104
32 libdispatch.dylib              0x64094 _dispatch_client_callout + 16
33 libdispatch.dylib              0x3f4a4 _dispatch_lane_serial_drain$VARIANT$armv81 + 600
34 libdispatch.dylib              0x3ff78 _dispatch_lane_invoke$VARIANT$armv81 + 440
35 libdispatch.dylib              0x3f380 _dispatch_lane_serial_drain$VARIANT$armv81 + 308
36 libdispatch.dylib              0x3ff44 _dispatch_lane_invoke$VARIANT$armv81 + 388
37 libdispatch.dylib              0x498e0 _dispatch_workloop_worker_thread + 608
38 libsystem_pthread.dylib        0x1e10 _pthread_wqthread + 284
39 libsystem_pthread.dylib        0x193c start_wqthread + 8
```
```
Thread
0  libsystem_kernel.dylib         0x1014 __workq_kernreturn + 8
1  libsystem_pthread.dylib        0x1e5c _pthread_wqthread + 360
2  libsystem_pthread.dylib        0x193c start_wqthread + 8
```
```
com.apple.uikit.eventfetch-thread
0  libsystem_kernel.dylib         0xaac mach_msg_trap + 8
1  libsystem_kernel.dylib         0x107c mach_msg + 72
2  CoreFoundation                 0x6768 __CFRunLoopServiceMachPort + 368
3  CoreFoundation                 0xaa70 __CFRunLoopRun + 1160
4  CoreFoundation                 0x1dc30 CFRunLoopRunSpecific + 572
5  Foundation                     0x17eac -[NSRunLoop(NSRunLoop) runMode:beforeDate:] + 232
6  Foundation                     0x56e90 -[NSRunLoop(NSRunLoop) runUntilDate:] + 88
7  UIKitCore                      0x4650a0 -[UIEventFetcher threadMain] + 512
8  Foundation                     0x64d2c __NSThread__start__ + 792
9  libsystem_pthread.dylib        0x3348 _pthread_start + 116
10 libsystem_pthread.dylib        0x1948 thread_start + 8
```
```
Thread
0  libsystem_kernel.dylib         0x1014 __workq_kernreturn + 8
1  libsystem_pthread.dylib        0x1e5c _pthread_wqthread + 360
2  libsystem_pthread.dylib        0x193c start_wqthread + 8
```
```
com.google.firebase.crashlytics.MachExceptionServer
0  libsystem_kernel.dylib         0xaac mach_msg_trap + 8
1  libsystem_kernel.dylib         0x107c mach_msg + 72
2  ClassDojo                      0xc1aad8 FIRCLSMachExceptionServer + 192 (FIRCLSMachException.c:192)
3  libsystem_pthread.dylib        0x3348 _pthread_start + 116
4  libsystem_pthread.dylib        0x1948 thread_start + 8
```
```
JavaScriptCore libpas scavenger
0  libsystem_kernel.dylib         0x1484 __psynch_cvwait + 8
1  libsystem_pthread.dylib        0x2860 _pthread_cond_wait$VARIANT$armv81 + 1224
2  JavaScriptCore                 0x1121410 scavenger_thread_main + 960
3  libsystem_pthread.dylib        0x3348 _pthread_start + 116
4  libsystem_pthread.dylib        0x1948 thread_start + 8
```
```
com.apple.NSURLConnectionLoader
0  libsystem_kernel.dylib         0xaac mach_msg_trap + 8
1  libsystem_kernel.dylib         0x107c mach_msg + 72
2  CoreFoundation                 0x6768 __CFRunLoopServiceMachPort + 368
3  CoreFoundation                 0xaa70 __CFRunLoopRun + 1160
4  CoreFoundation                 0x1dc30 CFRunLoopRunSpecific + 572
5  CFNetwork                      0x247dcc _CFURLStorageSessionDisableCache + 50388
6  Foundation                     0x64d2c __NSThread__start__ + 792
7  libsystem_pthread.dylib        0x3348 _pthread_start + 116
8  libsystem_pthread.dylib        0x1948 thread_start + 8
```

</details>

## Alamofire Environment
**Alamofire Version:** 5.6.1
**Dependency Manager:** SPM
**Xcode Version:** 13.3.1
**Swift Version:** 5.6
**Platform(s) Running Alamofire:** iOS
**macOS Version Running Xcode:** 12.3.1

**Comment by jshier** [comment]
This appears related to the update to use `SecTrustCopyCertificateChain` and the cast `as? [SecCertificate]`. https://github.com/Alamofire/Alamofire/blob/master/Source/ServerTrustEvaluation.swift#L611

Since it's a rare crash it seems like this approach generally works but there's some sort of implicit conversion under the hood that can fail. Can you copy past the text version of this crash (redacted if you want) into this thread so I can more easily pull the actual crash stack?

**Comment by Kaspik** [comment]
Thanks for quick reply! I can definitely copy paste here anything you need :)

```
Crashed: org.alamofire.session.rootQueue
0  libsystem_kernel.dylib         0x6bbc __pthread_kill + 8
1  libsystem_pthread.dylib        0xd854 pthread_kill + 208
2  libsystem_c.dylib              0x1f6a4 abort + 124
3  libswiftCore.dylib             0x2ed01c swift::warningv(unsigned int, char const*, char*) + 138
4  libswiftCore.dylib             0x2f6698 swift::MetadataAllocator::Allocate(unsigned long, unsigned long) + 516
5  libswiftCore.dylib             0x2f71f4 _swift_getGenericMetadata(swift::MetadataRequest, void const* const*, swift::TargetTypeContextDescriptor<swift::InProcess> const*) + 1004
6  libswiftCore.dylib             0x2dad68 __swift_instantiateCanonicalPrespecializedGenericMetadata + 32
7  libswiftFoundation.dylib       0x7a30 specialized _arrayConditionalCast<A, B>(_:) + 48
8  libswiftFoundation.dylib       0x1660c static Array._conditionallyBridgeFromObjectiveC(_:result:) + 80
9  ClassDojo                      0xf54fa0 AlamofireExtension<A>.certificates.getter + 611 (ServerTrustEvaluation.swift:611)
10 ClassDojo                      0xf54714 PinnedCertificatesTrustEvaluator.evaluate(_:forHost:) + 629 (ServerTrustEvaluation.swift:629)
11 ClassDojo                      0xf55110 protocol witness for ServerTrustEvaluating.evaluate(_:forHost:) in conformance PinnedCertificatesTrustEvaluator + 4354347280 (<compiler-generated>:4354347280)
12 ClassDojo                      0xf6ba6c SessionDelegate.attemptServerTrustAuthentication(with:) + 134 (SessionDelegate.swift:134)
13 ClassDojo                      0xf6e50c specialized SessionDelegate.urlSession(_:task:didReceive:completionHandler:) + 99 (<compiler-generated>:99)
14 ClassDojo                      0xf6d224 @objc SessionDelegate.urlSession(_:task:didReceive:completionHandler:) + 4354445860
15 CFNetwork                      0x7d898 _CFHostIsDomainTopLevelForCertificatePolicy + 10480
16 Foundation                     0x4e26c __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 16
17 Foundation                     0x5f920 -[NSBlockOperation main] + 100
18 Foundation                     0x39a8c __NSOPERATION_IS_INVOKING_MAIN__ + 20
19 Foundation                     0x49f20 -[NSOperation start] + 764
20 Foundation                     0x4d368 __NSOPERATIONQUEUE_IS_STARTING_AN_OPERATION__ + 20
21 Foundation                     0x5ab7c __NSOQSchedule_f + 180
22 libdispatch.dylib              0x46048 _dispatch_block_async_invoke2 + 104
23 libdispatch.dylib              0x64094 _dispatch_client_callout + 16
24 libdispatch.dylib              0x3f4a4 _dispatch_lane_serial_drain$VARIANT$armv81 + 600
25 libdispatch.dylib              0x3ff78 _dispatch_lane_invoke$VARIANT$armv81 + 440
26 libdispatch.dylib              0x3f380 _dispatch_lane_serial_drain$VARIANT$armv81 + 308
27 libdispatch.dylib              0x3ff44 _dispatch_lane_invoke$VARIANT$armv81 + 388
28 libdispatch.dylib              0x498e0 _dispatch_workloop_worker_thread + 608
29 libsystem_pthread.dylib        0x1e10 _pthread_wqthread + 284
30 libsystem_pthread.dylib        0x193c start_wqthread + 8
```

Do you need also other threads?

**Comment by jshier** [comment]
Can I get the exception information too?

**Comment by Kaspik** [comment]
Not sure what you mean by that?

But adding details of other thread just in case it helps too.

<img width="1242" alt="Screenshot 2022-05-16 at 20 33 51" src="https://user-images.githubusercontent.com/7204168/168659487-51bc159b-291d-4430-a800-028da895f46e.png">
<img width="1270" alt="Screenshot 2022-05-16 at 20 33 18" src="https://user-images.githubusercontent.com/7204168/168659510-8049e692-803e-4b9d-a3d0-a4e20ce8822e.png">
<img width="1272" alt="Screenshot 2022-05-16 at 20 33 03" src="https://user-images.githubusercontent.com/7204168/168659517-b442e3b2-e4f9-42e6-aa90-a2eb1ac34d4d.png">
<img width="1270" alt="Screenshot 2022-05-16 at 20 32 55" src="https://user-images.githubusercontent.com/7204168/168659531-5c31ab48-4443-4dee-8ced-42ee3cffb103.png">

**Comment by jshier** [comment]
Yeah, if I could get the text from those logs in the first image, that would be great. Contents of that trust thread are also interesting.

Basically, any info should be posted as text, not images, so I can copy / paste them elsewhere.

**Comment by Kaspik** [comment]
Got it, edited!

**Comment by jshier** [comment]
Thanks, this looks like this is the Swift runtime bug that affected older processors on 15.4.0 and was fixed in 15.4.1.

https://github.com/apple/swift/issues/58308

Thanks for the report details!
